### DISCUSSION TITLE
Workflow fails when Dask worker memory budget is exceeded

### DISCUSSION THREAD

**Discussion by FyzHsn** [AgnostiqHQ/covalent#1246] [Q&A] [answered]
My Machine Learning workflow execution fails during the preprocessing stage of the workflow. Having started the covalent server with `covalent start -d`, I was able to check the logs `covalent logs`. The log files show that the reason behind the workflow failure was:

```
2022-09-14 21:22:51,192 - distributed.worker_memory - WARNING - Worker tcp://127.0.0.1:64888 (pid=45315) exceeded 95% memory budget. Restarting...
```

It seems like the issue is with Dask worker memory being exceeded. I have two main questions:

1. How can I get more details on the Dask workers?
2. How can I move forward so that the workflow completes successfully?

**Comment by venkatBala** [comment, marked as answer]
The `Dask` dashboard will help get more details on the cluster workers. You will however need to install `bokeh`. The following steps would be helpful

* If covalent is already running, the server needs to be stopped
* `pip install bokeh`
* Start covalent
* Check the Dask dashboard at `localhost:8787` (In the workers tab you can inspect how much memory each Dask worker is running with)
* Secondly this information can also be retrieved via the `covalent cluster` CLI a.k.a `covalent cluster --info`

UI screenshots
![Screen Shot 2022-09-15 at 10 23 08 AM](https://user-images.githubusercontent.com/15014089/190430205-69cd73e7-2184-4c5f-b72f-b1561d310cee.png)
![Screen Shot 2022-09-15 at 10 23 27 AM](https://user-images.githubusercontent.com/15014089/190430220-9578f78b-f6aa-49cb-a5e0-d65611bb381c.png)

**Reply by FyzHsn** [reply]
In order to resolve the Dask worker memory issue, we can reduce the number of workers and increase the amount of memory allocated. Previously, I had 8 workers with 1 GB allocated to each.

To ensure that the Dask workers have enough memory to complete the task, I restarted covalent with:

```
covalent start -n 4 -m "2GB" -d
```

This allocated 2 GB per worker which turned out to be enough memory to complete the task.

**Reply by venkatBala** [reply]
Another alternative here would be to dispatch the compute intensive parts of the workflow to AWS via the Batch executor where memory is not a constraint per say
